rmw_cyclonedds_cpp/include/rmw_cyclonedds_cpp/serdes.hpp

```diff
@@ -35,122 +35,62 @@ class cycser
-  inline cycser & operator<<(int8_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(uint8_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(int16_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(uint16_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(int32_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(uint32_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(int64_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(uint64_t x) {serialize(x); return *this;}
-  inline cycser & operator<<(float x) {serialize(x); return *this;}
-  inline cycser & operator<<(double x) {serialize(x); return *this;}
-  inline cycser & operator<<(const char * x) {serialize(x); return *this;}
-  inline cycser & operator<<(const std::string & x) {serialize(x); return *this;}
-  inline cycser & operator<<(const std::wstring & x) {serialize(x); return *this;}
-  template<class T>
-  inline cycser & operator<<(const std::vector<T> & x) {serialize(x); return *this;}
-  template<class T, size_t S>
-  inline cycser & operator<<(const std::array<T, S> & x) {serialize(x); return *this;}
+  template <typename T>
+  cycser & operator<<(const T &x) {serialize(x); return *this;}
 
-#define SER(T) inline void serialize(T x) { \
-    if ((off % sizeof(T)) != 0) { \
-      off += sizeof(T) - (off % sizeof(T)); \
-    } \
-    resize(off + sizeof(T)); \
-    *(reinterpret_cast<T *>(data() + off)) = x; \
-    off += sizeof(T); \
-}
-  SER(char);
-  SER(int8_t);
-  SER(uint8_t);
-  SER(int16_t);
-  SER(uint16_t);
-  SER(int32_t);
-  SER(uint32_t);
-  SER(int64_t);
-  SER(uint64_t);
-  SER(float);
-  SER(double);
-#undef SER
+  template <typename T>
+  void serialize(T x) {
+    if ((off % sizeof(T)) != 0) {
+      off += sizeof(T) - (off % sizeof(T));
+    }
+    resize(off + sizeof(T));
+    *(reinterpret_cast<T *>(data() + off)) = x;
+    off += sizeof(T);
+  }
 
   inline void serialize(bool x)
   {
-    serialize(static_cast<unsigned char>(x));
-  }
-  inline void serialize(const char * x)
-  {
-    size_t sz = strlen(x) + 1;
-    serialize(static_cast<uint32_t>(sz));
-    resize(off + sz);
-    memcpy(data() + off, x, sz);
-    off += sz;
+      serialize(static_cast<unsigned char>(x));
   }
+
   inline void serialize(const std::string & x)
   {
-    size_t sz = x.size() + 1;
-    serialize(static_cast<uint32_t>(sz));
-    resize(off + sz);
-    memcpy(data() + off, x.c_str(), sz);
-    off += sz;
+      serialize_list(x.c_str(), x.size()+1);
   }
+
   inline void serialize(const std::wstring & x)
   {
-    size_t sz = x.size();
-    serialize(static_cast<uint32_t>(sz));
-    resize(off + sz * sizeof(wchar_t));
-    memcpy(data() + off, reinterpret_cast<const char *>(x.c_str()), sz * sizeof(wchar_t));
-    off += sz * sizeof(wchar_t);
+      serialize_list(x.c_str(), x.size()); // todo: is the size right or do we need to include the null terminator?
```

Comment:
@eboasson What's the intent here?
(1) should the string be both prefixed by a count *and* null-terminated?
(2) should the serialized wstring count include the null character?

Reply:
The CDR spec is very clear about ordinary strings: it must be null-terminated and the count must include the terminator (so the count for an empty string must be 1). For wide strings I simply matched what Fast-RTPS did. (Neither OpenSplice nor Cyclone supports wide strings natively yet, thankfully! But it also means I only have the specs and Fast-RTPS ...)

The specs are a bit of a mess with wide strings ([CORBA interoperability](https://www.omg.org/spec/CORBA/ISO/19500-2/PDF), section 9.4.2.7 is one; [XTypes](https://www.omg.org/spec/DDS-XTypes/1.3/Beta1/PDF) is another) and with the two together I think there is only a single interpretation:

Section 9.4.2.7 on wide strings:

> For GIOP version 1.1, a wide string is encoded as an unsigned long indicating the length of the string in octets or unsigned integers (determined by the transfer syntax for wchar) followed by the individual wide characters. The string contents include a single terminating null character. The string length includes the null character. The terminating null character for a wstring is also a wide character.
> For GIOP version 1.2 and 1.3, when encoding a wstring, always encode the length as the total number of octets used by the encoded value, regardless of whether the encoding is byte-oriented or not. For GIOP version 1.2 and 1.3 a wstring is not terminated by a null character. In particular, in GIOP version 1.2 and 1.3 a length of 0 is legal for wstring.
> NOTE: For GIOP versions prior to 1.2 and 1.3, interoperability for wstring is limited to the use of two-octet fixed-length encoding.
> Wstring values in encapsulations are assumed to be encoded using GIOP version 1.2 and 1.3 CDR.

GIOP versions go up to at least 1.4, which I guess means one should follow GIOP 1.3? And "encapsulations" are things in themselves. The DDS XTypes spec 1.3 (which is not even adopted yet) cleans things up a bit, and it says:

> 7.3.1.7 String Types
> The string container defined by this specification has two element types for which the behavior is defined: Char8 and Char16. Strings of Char8 shall be represented by the IDL type string. Strings of Char16 shall be represented by the IDL type wstring. In either case, any bound shall be retained.

So using UTF-32 (which is what wchar_t typically is) is not a smart choice. Methinks the length should be in octets and a terminating 0 is not supposed to be present.

Incidentally, wide characters suffer from similar problems. I'm not even sure whether an IDL wchar should be a code point or a grapheme ... If a code point, UTF-32 for a character (with endianness determined by the context) makes, if a grapheme, the special-cased sequence-of-octets encoding that section 9.4.1.6 uses makes more sense.

My apologies for also being confused by the situation!

Reply:
Oh wow. That is confusing! According to the footnote on this page, CORBA is not a relevant standard for DDS. Also, GIOP is not mentioned on this page at all: https://www.dds-foundation.org/omg-dds-standard/

> DDS™ and CORBA® are two independent connectivity standards ... They both provide application portability (using standard language interfaces) and network interoperability. Both DDS and CORBA were developed by the Object Management Group® (OMG). However, they were designed to support different use cases and integration patterns. ... While the DDS and CORBA standards are independent, both leverage the OMG Interface Definition Language™ (IDL™) to specify data and service types.

It very clearly says XTypes is the relevant serialization format:

> DDS-XTypes v1.2 – defines the DDS Type System and the serialized representation of DDS data.

Here's what XTypes says about strings, which I think we do a poor job of following:

> Objects of `String<Char8>` type shall be represented using the UTF-8 character encoding. The serialized length of an object of type String<Char8> shall be the number of bytes in the CDR buffer taken by the `String<Char8>` characters, including the terminating NUL character. The serialized length may not be the same as the number of Unicode characters because a single Unicode character encoded using the UTF-8 encoding may take one to four bytes.

> Objects of `String<Char16>` type shall be represented using the UTF-16 character encoding. The serialized length of an object of type `String<Char16>` shall be the number of bytes in the CDR buffer taken by the `String<Char16>` characters. This is twice the number of characters in the string because a single character (in the Basic Multilingual Plane) encoded using UTF-16 takes 2 bytes to serialize.

> The UTF-16 representation of object of type `String<Char16>` shall not include a Byte Order Mark (BOM). The representation shall also not include any terminating NUL character(s).

Reply:
That was quite a rabbit hole...

Reply:
Quite the rabbit hole indeed! I'm grateful you went down it, because I had completely forgotten that XTypes nowadays overrides the older ones, and so I would probably have spent some time looking in the wrong places ... (In the olden days, before XTypes, it just referred to CDR, and that GIOP spec is the only definition I have ever been able to find for CDR. Whether everything's properly compatible with that is an interesting but mostly academic question that I'm not inclined to study now.)

My sense of things is that the XTypes simply standardized RTI's practice. Which is fine, because at least there is a specification now. And for sure, we do a poor job of following it.

I think the only strategy that is both reasonable and viable long-term is to follow the XTypes spec.